docs/architecture.md

```diff
@@ -0,0 +1,221 @@
+the runtime state of CRI-RM across restarts.
+
+The cache provides functions for querying and updating the state of pods and
+containers. This is the mechanism used by the active policy to make resource
+assignment decisions. The policy simply updates the state of the affected
+containers in the cache according to the decisions.
+
+The cache's ability to associate and track changes to containers with resource
+domains is used to enforce policy decisions. The generic controller layer first
+queries which containers have pending changes, then invokes each controller for
+each containers. The controllers use the querying functions provided by the
+cache to decide if anything in their resource/control domain needs to be changed
+and then act accordingly.
+
+Access to the cache needs to be serialized. However, this serialization is not
+provided by the cache itself. Instead, it assumes callers to make sure proper
+protection is in place against concurrent read-write access. The request and
+event processing pipelines in the resource manager use a lock to serialize
+request and event processing and consequently access to the cache.
+
+
+### Generic Policy Layer
+
+The generic policy layer defines the abstract interface the rest of CRI-RM
+uses to interact with policy implementations and takes care of the details
+of activating and dispatching calls through to the configured active policy.
+
+
+### Generic Resource Controller Layer
+
+The generic resource controller layer defines the abstract interface the rest
+of CRI-RM uses to interact with resource controller implementations and takes
+care of the details of dispatching calls to the controller implementations for
+post-policy enforcment of decisions.
+
+
+### Metrics Collector
+
+
+
+
+
+### Policy Implementations
+
+#### None
+
+An empty policy which does no policy decisions. It is included merely for
+the sake of completeness, analoguous to the none policy of the CPU Manager
+in kubelet.
+
+#### Static Pools
+
+A backward-compatible reimplementation of [CMK](https://github.com/intel/CPU-Manager-for-Kubernetes)
+for CRI-RM with a few extra features.
+
+#### Static
+
+Part of the code from the static policy of CPU Manager in kubelet, brutally
+hacked to work within CRI-RM. Serves merely as a proof-of-concept that the
+current policies of kubelet can be implemented in CRI-RM.
+
+#### Static Plus
+
+A fairly simplistic policy similar in spirit to the static policy of
+CPU Manager in kubelet, with a few extra features. 
+
+#### Memtier
+
+A topology-aware policy capable of handling multiple tiers/types of memory,
+typically a DRAM/PMEM combination configured in 2-layer memory mode.
+
+### Resource Controller Implementations
+
+#### RDT
+
+A resource controller implementation responsible for enforcing assignments
+of controllers to last level cache and memory bandwidth allocation classes,
+using the resctrl pseudo filesystem of the Linux kernel.
```

Comment:
Could we simplify wording here... Something like

> RDT controls containers' last level cache allocations and memory bandwith using resctrl pseudo filesystem on Linux.

Then maybe a separate sentence or two on how input is structured and how it maps these resources to pods and containers.

Reply:
I tried to reformulate the descriptions of controllers, trying to avoid too long sentences.